render/egl.c

```diff
@@ -95,7 +95,7 @@ static void print_dmabuf_formats(struct wlr_egl *egl) {
 		return;
 	}
 
-	char *str_formats = malloc(num * 5);
+	char *str_formats = malloc((num * 5) + 1);
 	if (str_formats == NULL) {
```

Comment:
(latest commit) I'd rather change the rest -- either return early from the function or check to `str_formats == NULL && num != 0` -- bailing out early or `foo = malloc(0); ...that doesn't use foo..; free(foo);`

Reply:
Could you clarify for me please?

Reply:
Sorry, I thought that +1 was the same as an earlier comment because of the commit message e.g. "the VLA code used +1 because a VLA of zero size is undefined behaviour, while it's perfectly OK to do malloc(0)" -- but it's actually needed for the terminal nul byte.

This one is fine, but if there are other malloc(x+1) where x can be 0 because of the VLa undefined behaviour, we don't really need that +1 anymore, it's ok to malloc(0) -- that'll return NULL or some memory pointer you're not supposed to use depending on the system, but it's perfectly fine as long as it's not accessed. Sorry for not taking the time to do a proper review.


That being said, this code has two problems:
 - If you add a malloc, you always need to add a free somewhere. I didn't pay attention  earlier but as it is this `str_formats` is leaking.
 - if num is 0, nothing prints the terminal nul byte for wlr_log to be safe - it will likely print garbage after the allocated region. You need to at least set the last byte of str_formats to 0 or '\0' (or use calloc).

Reply:
This is very helpful, thank you. `str_formats` was indeed leaking and I've now corrected that. I've done my best to remove the unnecessary `+ 1`s which were avoiding the UB you mentioned. To fix the `num == 0` issue, I simply guarded the code with a condition which tests for `num > 0`. It looks like the code didn't need to execute if `num == 0`.